### PR TITLE
[SPARK-38524][SPARK-38553][K8S] Bump `Volcano` to v1.5.1 and fix Volcano `weight` to be positive integer and use `cpu` capability instead

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
- Bump Volcano to v1.5.1: https://github.com/volcano-sh/volcano/issues/2090
- Use capability to limit disable queue

### Why are the changes needed?
In Volcano, weight should be [a positive integer](https://github.com/volcano-sh/volcano/blob/c5bed286211af4f9d94503968c1c6b06baa882d1/pkg/webhooks/admission/queues/validate/validate_queue.go#L188), so weight 0 is a wrong usage. As [description](https://volcano.sh/en/docs/queue/) for queue
- weight is a soft constraint.
- capability is a hard constraint.

We better to use capability to limit disable queue. This also fix the error `requestBody.spec.weight: Invalid value: 0: queue weight must be a positive integer` when running latest volcano image.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
```
# arm64
kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/v1.5.1/installer/volcano-development-arm64.yaml

# x86
kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/v1.5.1/installer/volcano-development.yaml

build/sbt -Pvolcano -Pkubernetes -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube,r  -Dtest.include.tags=volcano  -Dspark.kubernetes.test.namespace=default "kubernetes-integration-tests/testOnly"
```

### CODE REVIEW

**Comment by Yikun** [comment]
cc @dongjoon-hyun

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
What is the Volcano versioning? I don't believe this is sustainable to depend on `latest`. Please be specific about 
- which Volcano version has these enforcements
- which Volcano version can delete the `Open` queue.

Without a specific volcano version, everything is meaningless because it's fragile.

**Comment by Yikun** [comment]
Sorry for the trouble.

> What is the Volcano versioning?

- Volcano release docker image with specific tag, such as v1.5.0, next release would be v1.5.1
- Fixes will merge to latest and then backport to v1.5.x.

> I don't believe this is sustainable to depend on latest
- We use master latest images for spark IT test before. **I have to say sorry about this, I also think it's my fault to make this decision**, I just considered it's also very import for us to do continuesly validation but ignored the stable factor. 
- We need to use I also track issue on: https://github.com/volcano-sh/volcano/issues/2076 to release v1.5.1 beta release for continuesly validation, this issue will be completed today.
- These also expose some shortcomings of Volcano in dev versioning in cross community, but the Volcano community is also actively improving, hope you can understand.

> which Volcano version has these enforcements?

Always, but there was a compatiable problem in v1.5.0 webhook feature (admissionconrol, input parameter validation) are not work in v1.5.0, so we didn't see this enforcements. Epecially in kubernetes v1.22+ because of [admissionregistration.k8s.io/v1beta1](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#webhook-resources-v122) drop, alread be fixed in volcano master https://github.com/volcano-sh/volcano/issues/2063 and will be backported to v1.5.x soon.

> which Volcano version can delete the Open queue?

Volcano can delete the Open queue improved since v1.5.1.beta.0 until queue close API supported for ease of use, volcano will also supported first close and delete queue api in v1.6.

TLDR:
- https://github.com/volcano-sh/volcano/issues/2076 volcano is going to release a v1.5.1-beta.0 for continuesly validation, we still can use release-1.5/installer/volcano-development-*.yaml to do validation as .
- Then I will do a validation on with volcano `v1.5.1-beta.0` make sure all volcano feature works before 15.March.
- This PR will be mark as ready for review soon. : )

**Comment by dongjoon-hyun** [comment]
Thank you for the detail, @Yikun !

**Comment by dongjoon-hyun** [comment]
Hi, @Yikun . Is there a pre-defined release cadence in Volcano community?

**Comment by Yikun** [comment]
@dongjoon-hyun Thanks for notification, the beta image will be published in 1-2 hours. I also do a pre-validation on x86 and arm based on volcano latest branch, all IT passed as expected with this PR. I will ping you once it's ready.

@Thor-wl @william-wang is doing final check and then release the beta image, thank for help and efforts at the weekend.

**Comment by dongjoon-hyun** [comment]
Thank you for the confirmation. Please note that Apache Spark cannot depend on `SNAPSHOT` or `beta` stuff.

**Comment by Yikun** [comment]
@dongjoon-hyun Yes for sure, the target version would be `v1.5.1`.

As we dicussed before, the release time would be a time after Spark 3.3 branch cut down (15. March) before Spark 3.3 release RC (April).

**Comment by Yikun** [comment]
All test passed with volcano release-1.5 (v1.5.1-beta.0 images) on x86 and arm64.

<details><summary>test on arm64 details</summary>

```
# arm64
kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/release-1.5/installer/volcano-development-arm64.yaml

build/sbt -Pvolcano -Pkubernetes -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube,r  -Dtest.include.tags=volcano  -Dspark.kubernetes.test.namespace=default "kubernetes-integration-tests/testOnly"

[info] VolcanoSuite:
[info] - Run SparkPi with no resources (12 seconds, 370 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (11 seconds, 520 milliseconds)
[info] - Run SparkPi with a very long application name. (11 seconds, 927 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (11 seconds, 1 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (10 seconds, 899 milliseconds)
[info] - Run SparkPi with an argument. (11 seconds, 943 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (11 seconds, 979 milliseconds)
[info] - All pods have the same service account by default (10 seconds, 962 milliseconds)
[info] - Run extraJVMOptions check on driver (6 seconds, 105 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (12 seconds, 17 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (19 seconds, 699 milliseconds)
[info] - Run SparkPi with env and mount secrets. (22 seconds, 314 milliseconds)
[info] - Run PySpark on simple pi.py example (13 seconds, 57 milliseconds)
[info] - Run PySpark to test a pyfiles example (15 seconds, 144 milliseconds)
[info] - Run PySpark with memory customization (12 seconds, 943 milliseconds)
[info] - Run in client mode. (8 seconds, 259 milliseconds)
[info] - Start pod creation from template (13 seconds, 95 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (11 seconds, 987 milliseconds)
[info] - Test basic decommissioning (46 seconds, 368 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (46 seconds, 699 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 46 seconds)
[info] - Test decommissioning timeouts (47 seconds, 204 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 8 seconds)
[info] - Run SparkPi with volcano scheduler (13 seconds, 27 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minCPU (38 seconds, 825 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minMemory (38 seconds, 911 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (19 seconds, 139 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (34 seconds, 263 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (18 seconds, 491 milliseconds)
[info] Run completed in 13 minutes, 16 seconds.
[info] Total number of tests run: 29
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 29, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 829 s (13:49), completed 2022-3-13 20:29:27
```
</details>

<details><summary>test on x86 details </summary>

```
# x86
kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/release-1.5/installer/volcano-development.yaml

build/sbt -Pvolcano -Pkubernetes -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube,r  -Dtest.include.tags=volcano  -Dspark.kubernetes.test.namespace=default "kubernetes-integration-tests/testOnly"

[info] VolcanoSuite:
[info] - Run SparkPi with no resources (12 seconds, 171 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (11 seconds, 741 milliseconds)
[info] - Run SparkPi with a very long application name. (12 seconds, 794 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (11 seconds, 740 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (11 seconds, 711 milliseconds)
[info] - Run SparkPi with an argument. (12 seconds, 789 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (11 seconds, 848 milliseconds)
[info] - All pods have the same service account by default (12 seconds, 756 milliseconds)
[info] - Run extraJVMOptions check on driver (6 seconds, 661 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (12 seconds, 881 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (19 seconds, 192 milliseconds)
[info] - Run SparkPi with env and mount secrets. (20 seconds, 721 milliseconds)
[info] - Run PySpark on simple pi.py example (13 seconds, 776 milliseconds)
[info] - Run PySpark to test a pyfiles example (15 seconds, 767 milliseconds)
[info] - Run PySpark with memory customization (13 seconds, 738 milliseconds)
[info] - Run in client mode. (9 seconds, 176 milliseconds)
[info] - Start pod creation from template (11 seconds, 792 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (12 seconds, 831 milliseconds)
[info] - Test basic decommissioning (47 seconds, 65 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (48 seconds, 200 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 46 seconds)
[info] - Test decommissioning timeouts (47 seconds, 423 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 10 seconds)
[info] - Run SparkPi with volcano scheduler (11 seconds, 695 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minCPU (34 seconds, 328 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minMemory (32 seconds, 235 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (18 seconds, 230 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (29 seconds, 370 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (19 seconds, 213 milliseconds)
[info] Run completed in 13 minutes, 13 seconds.
[info] Total number of tests run: 29
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 29, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 837 s (13:57), completed Mar 13, 2022 8:29:45 PM
```
</details>

Ready to go!

**Review by martin-g — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
> @dongjoon-hyun Yes for sure, the target version would be `v1.5.1`.
> 
> As we dicussed before, the release time would be a time after Spark 3.3 branch cut down (15. March) before Spark 3.3 release RC (April).

Yes, For the test PR, it sounds good because March 15 is a deadline for `Feature Freeze`.

File: resource-managers/kubernetes/integration-tests/src/test/resources/volcano/disable-queue.yml [hunk 1]
```diff
@@ -19,6 +19,6 @@ kind: Queue
 metadata:
   name: queue
 spec:
-  weight: 0
+  weight: 1
   capability:
-    cpu: "1"
+    cpu: "0.1"
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/resources/volcano/disable-queue.yml:24]: `cpu: 0` is also invalid by definition?

Comment (Yikun) [reply resource-managers/kubernetes/integration-tests/src/test/resources/volcano/disable-queue.yml:24]: cap 0 is valid, if capacity <= 0 represents [no limit](https://github.com/volcano-sh/volcano/blob/6ea23412696eb7369e8e3dcf543604a5341058e1/pkg/scheduler/plugins/proportion/proportion.go#L104-L112) in Volcano queue.

This is different with `Weight`, weight is soft constraint to calculate some proportions according to weight value, so `0` bring some unexpected behavior, so it's invalid.

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/integration-tests/src/test/resources/volcano/disable-queue.yml:24]: Got it. If then, could you use more smaller value like `0.0001` than `0.1`?

Comment (Yikun) [reply resource-managers/kubernetes/integration-tests/src/test/resources/volcano/disable-queue.yml]: Sure, unit of least precision is 0.001 as kuberentes. So I just set to `0.001`.
 
PR updated.

**Comment by Yikun** [comment]
```
[info] VolcanoSuite:
[info] - Run SparkPi with volcano scheduler (11 seconds, 670 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minCPU (33 seconds, 564 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minMemory (32 seconds, 829 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (15 seconds, 918 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (31 seconds, 833 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (18 seconds, 571 milliseconds)
[info] Run completed in 2 minutes, 29 seconds.
[info] Total number of tests run: 6
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 6, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 303 s (05:03), completed 2022-3-15 11:28:21
```

**Review by martin-g — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
It seems that `v1.5.1-beta.0` has another installation issue.
```
          image: volcanosh/vc-controller-manager-arm64:v1.5.1-beta.0
...
      message: 'Internal error occurred: failed calling webhook "mutatepod.volcano.sh":
        Post "https://volcano-admission-service.volcano-system.svc:443/pods/mutate?timeout=10s":
        no endpoints available for service "volcano-admission-service"'
```

**Comment by dongjoon-hyun** [comment]
I'm not sure what is going on `Volcano` project, but volcano installation and cleanup seems incomplete.
It screwed the EKS cluster itself.

```
22/03/14 23:32:49 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
22/03/14 23:32:49 INFO SparkKubernetesClientFactory: Auto-configuring K8S client using current context from users K8S config file
22/03/14 23:32:50 INFO KerberosConfDriverFeatureStep: You have not specified a krb5.conf file locally or via a ConfigMap. Make sure that you have the krb5.conf locally on the driver image.
22/03/14 23:32:57 ERROR Client: Please check "kubectl auth can-i create pod" first. It should be yes.
Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: .... Message: 
Internal error occurred: failed calling webhook "mutatepod.volcano.sh": Post "https://volcano-admission-service.volcano-system.svc:443/pods/mutate?timeout=10s": service "volcano-admission-service" not found. Received status: Status(apiVersion=v1, code=500, details=StatusDetails(causes=[StatusCause(field=null, message=failed calling webhook "mutatepod.volcano.sh": Post "https://volcano-admission-service.volcano-system.svc:443/pods/mutate?timeout=10s": service "volcano-admission-service" not found, reason=null, additionalProperties={})], group=null, kind=null, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Internal error occurred: failed calling webhook "mutatepod.volcano.sh": Post "https://volcano-admission-service.volcano-system.svc:443/pods/mutate?timeout=10s": service "volcano-admission-service" not found, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=InternalError, status=Failure, additionalProperties={}).
```

**Comment by dongjoon-hyun** [comment]
It was my bad. I made a follow-up to fix it.
- https://github.com/apache/spark/pull/35859

**Comment by Yikun** [comment]
@dongjoon-hyun See this: https://github.com/volcano-sh/volcano/issues/2079, note that how to cleanup volcano completely

**Comment by Yikun** [comment]
Then after above clean operations, you can try just use `release-1.5` (`v1.5.1-beta.0`) to test.

**Comment by dongjoon-hyun** [comment]
@Yikun . I cannot verify your PR because surprisingly `Volcano v1.5.0` fails to install on a new vanilla EKS cluster from the start.

```
$ k apply -f https://raw.githubusercontent.com/volcano-sh/volcano/v1.5.0/installer/volcano-development-arm64.yaml
namespace/volcano-system created
configmap/volcano-scheduler-configmap created
serviceaccount/volcano-scheduler created
clusterrole.rbac.authorization.k8s.io/volcano-scheduler created
clusterrolebinding.rbac.authorization.k8s.io/volcano-scheduler-role created
deployment.apps/volcano-scheduler created
serviceaccount/volcano-admission created
clusterrole.rbac.authorization.k8s.io/volcano-admission created
clusterrolebinding.rbac.authorization.k8s.io/volcano-admission-role created
deployment.apps/volcano-admission created
service/volcano-admission-service created
job.batch/volcano-admission-init created
serviceaccount/volcano-controllers created
clusterrole.rbac.authorization.k8s.io/volcano-controllers created
clusterrolebinding.rbac.authorization.k8s.io/volcano-controllers-role created
deployment.apps/volcano-controllers created
error: error validating "https://raw.githubusercontent.com/volcano-sh/volcano/v1.5.0/installer/volcano-development-arm64.yaml": error validating data: [ValidationError(CustomResourceDefinition.spec): unknown field "subresources" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionSpec, ValidationError(CustomResourceDefinition.spec): unknown field "validation" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionSpec, ValidationError(CustomResourceDefinition.spec): unknown field "version" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionSpec, ValidationError(CustomResourceDefinition.spec): missing required field "versions" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.CustomResourceDefinitionSpec]; if you choose to ignore these errors, turn validation off with --validate=false
```

**Comment by dongjoon-hyun** [comment]
I'm trying to rebuilt the cluster. I'll let you know if I succeed again.

**Comment by Yikun** [comment]
```
k delete validatingwebhookconfigurations volcano-admission-service-jobs-validate volcano-admission-service-pods-validate volcano-admission-service-queues-validate
k delete mutatingwebhookconfigurations volcano-admission-service-jobs-mutate volcano-admission-service-podgroups-mutate volcano-admission-service-pods-mutate volcano-admission-service-queues-mutate
# Using beta
kubectl apply -f https://raw.githubusercontent.com/volcano-sh/volcano/release-1.5/installer/volcano-development-arm64.yaml
```

Try this, as I mentioned before, there are some bugs fixed after volcano v1.5.0 release. such as:

- https://github.com/volcano-sh/volcano/commit/42fd4883189e47d2555f71b26182ae5e13651931 arm64 deploy
- https://github.com/volcano-sh/volcano/commit/7a5d8be276254aaec1d662fc1998a90f8b7e2b68 enable webhook
- https://github.com/volcano-sh/volcano/commit/54446650eca749594fc21949223c14fb7cabc8de support queue deleted by kubectl

or you mean we have to first bump volcano to v1.5.1, then merge this PR?

**Comment by dongjoon-hyun** [comment]
Ya, I found that I was hitting the original issue of lack of ARM64 support of `Volcano`.
- https://github.com/volcano-sh/volcano/commit/42fd4883189e47d2555f71b26182ae5e13651931

**Comment by Yikun** [comment]
>  I found that I was hitting the original issue of lack of ARM64 support of Volcano

So my suggestion is first we do validation on `release-1.5`, if all is ready, volcano will realese v1.5.1 as official first support. WDYT?

**Comment by dongjoon-hyun** [comment]
I don't think that's Apache Spark's role. Apache Spark community will mark this is a known issue in the release note if Volcano doesn't provide any release until our release date, @Yikun .

**Comment by Yikun** [comment]
@dongjoon-hyun OK, so now we have only one way to solve this issue: volcano release v1.5.1 before cut down?

**Comment by dongjoon-hyun** [comment]
Unfortunately, I don't know what is inside Volcano code, @Yikun . It's a role of Volcano community to provide a reliable tested release. Given the current circumstance I observed so far during helping this area, I don't know `v1.5.1` will be healthy or not.

We will test the stability again when it's available publicly.

**Comment by Yikun** [comment]
@dongjoon-hyun OK, thanks for your suggestion, I will discussion with Volcano community! Thanks!

**Comment by Yikun** [comment]
@dongjoon-hyun  I just talked with Volcano group member, then get all vote +1 for v1.5.1 (same code base with v1.5.1.beta.0). The release will ready in next 2 hours. https://github.com/volcano-sh/volcano/issues/2090

I also updated this PR to bump v1.5.1 and fix the test issue, I could separate version bump and test case, so just squash them in to this PR.

**Comment by Yikun** [comment]
@dongjoon-hyun  Volcano are considering to release v1.5.1 (same code base with v1.5.1.beta.0). The release will ready soon. https://github.com/volcano-sh/volcano/issues/2090

I also updated this PR to bump v1.5.1 and fix the test issue, I couldn't separate version bump and test case, so just squash them in to this PR.

**Comment by Yikun** [comment]
Ready for review.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. It's great, @Yikun .
Merged to master.
